build-scripts/generateConfigurationAttributes.ts

```diff
@@ -0,0 +1,60 @@
+/*!
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/* tslint:disable */
+
+import * as fs from 'fs'
+// tslint:disable-next-line:no-implicit-dependencies
+import { JSONSchema4 } from 'json-schema'
+import { compile } from 'json-schema-to-typescript'
+import * as packageJson from '../package.json'
+
+const config = [
+    {
+        debugger: 'aws-sam',
+        requestType: 'direct-invoke' as 'direct-invoke',
+        outputFile: 'src/shared/sam/debugger/awsSamDebugConfiguration.gen.ts',
+        imports: ["import * as vscode from 'vscode'"],
+        topLevelClass: 'AwsSamDebuggerConfiguration'
+    }
+]
+
+const header = `
+/* tslint:disable */
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/**
+ * This file was automatically generated by json-schema-to-typescript.
+ * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
+ * and run json-schema-to-typescript to regenerate this file.
+ */
+
+import * as vscode from 'vscode'
+`
+
+async function generateConfigurationAttributes(): Promise<void> {
+    for (const debugConfiguration of packageJson.contributes.debuggers) {
+        const debuggerConfig = config.find(cfg => {
+            return cfg.debugger === debugConfiguration.type
+        })
+        if (debuggerConfig === undefined) {
+            continue
+        }
+        const schema = debugConfiguration.configurationAttributes[debuggerConfig.requestType] as JSONSchema4
+        await compile(schema, 'DirectInvoke', { bannerComment: header })
+            .then(ts => addBaseClass(ts, debuggerConfig.topLevelClass))
```

Comment:
Since this is being awaited, I don't think promise chaining is needed.

```
const ts = await compile(schema, 'DirectInvoke', { bannerComment: header })
addBaseClass(ts, debuggerConfig.topLevelClass)
fs.writeFileSync(debuggerConfig.outputFile, ts)
```

Reply:
@ckant 's suggestion has a bug -- you need the retval from addBaseClass. Otherwise, absolutely chaining can be avoided.

Reply:
Thanks, fixed

Reply:
I'm a fan of functional style chaining over intermediate objects. To make it clearer I could rename the `ts` arguments to something that conveys more information if this is unclear to people.

Reply:
small scopes should not require descriptive names. This scope is as small as it gets.

> chaining over intermediate objects

+1, 

- it conveys the dependency chain without the reader needing to piece that together. 
- it makes it easy to see that the variables aren't used elsewhere in the scope

 question is whether this handles exceptions the same way as sequential `await` would?

Reply:
I agree in general for chaining async stuff. It's just that the first call to `then` is returning the (sync) result of `addBaseClass` wrapped in a promise which makes it look async.

Reply:
Versus the suggestion which makes it explicit that the first call is async and the others are not. Not a big deal, though.

Reply:
the net impact is that the whole thing is async tho  -- so -- does it matter? I'd argue no.

Reply:
Yeah it's just similar to that [function coloring](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) thing and the stuff that kotlin was trying to get away from with coroutines. Async stuff requiring its own ceremony which ends up changing the way you invoke functions. Without promises/chaining, you would have written these as separate statements.

Result is the same though for sure.

Reply:
> Without promises/chaining, you would have written these as separate statements

Why? In "true" functional languages such as haskell,  this is the "monad" pattern. Similar to a "pipeline". 

IMO this is a serendipitous effect of `then()` (though not the main purpose of course): it makes the relationship clear.

Reply:
Right, I'm not trying to argue against the functional style. I prefer it, in general, as well. The argument was that the Promise chaining hides the sync operations and results in different-looking code.

There's [a proposal for a pipeline operator in JS](https://github.com/tc39/proposal-pipeline-operator). If implemented (and idiomatic), that would give us the best of both worlds here. It would make `await`ing explicit and reduce verbosity and scope pollution brought about by the intermediate statements.

Reply:
> There's [a proposal for a pipeline operator in JS](https://github.com/tc39/proposal-pipeline-operator). 

Interesting! Though that looks closer to [UFCS](https://en.wikipedia.org/wiki/Uniform_Function_Call_Syntax).